pkg/controller/monitor/cnr_indicator.go

```diff
@@ -0,0 +1,211 @@
+)
+
+const (
+	metricsNameCNRReportAnomaly  = "cnr_report_anomaly"
+	metricsNameCNRReportLantency = "cnr_report_latency"
+)
+
+const (
+	// reasonNumaExclusiveAnomaly is the reason for numa exclusive anomaly
+	// when numa_binding and numa_exclusive are both set
+	// the pod with numa_binding and numa_exclusive shares the numa with other pods
+	reasonNumaExclusiveAnomaly = "The pod with numa_binding and numa_exclusive shares the numa with other pods"
+	// reasonNumaAllocatableSumAnomaly is the reason for numa allocatable sum anomaly
+	// when the node's sum of numa allocatable is not equal to the node allocatable
+	reasonNumaAllocatableSumAnomaly = "The node's sum of numa allocatable is not equal to the node allocatable"
+	// reasonPodAllocationSumAnomaly is the reason for pod allocation sum anomaly
+	// when the numa's sum of pod allocation is greater than the numa allocatable
+	reasonPodAllocationSumAnomaly = "The numa's sum of pod allocation is greater than the numa allocatable"
+)
+
+// checkNumaExclusiveAnomaly checks whether exist the pod with numa_binding and numa_exclusive shares the numa with other pods
+func (ctrl *CNRMonitorController) checkNumaExclusiveAnomaly(cnr *v1alpha1.CustomNodeResource) bool {
+	qosConf := generic.NewQoSConfiguration()
+	for _, socket := range cnr.Status.TopologyZone {
+		for _, numa := range socket.Children {
+			if numa.Type != v1alpha1.TopologyTypeNuma {
+				// only check numa
+				continue
+			}
+			podNum := len(numa.Allocations)
+			for _, allocation := range numa.Allocations {
+				key := allocation.Consumer
+				namespace, podname, _, err := native.ParseUniqObjectUIDKey(key)
+				if err != nil {
+					klog.Errorf("[CNRIndicatorNumaExclusiveAnomaly] failed to parse uniq object uid key %s", key)
+					continue
+				}
+				pod, err := ctrl.podLister.Pods(namespace).Get(podname)
+				if err != nil {
+					klog.Errorf("[CNRIndicatorNumaExclusiveAnomaly] failed to get pod %s", key)
+					continue
+				}
+				if qos.IsPodNumaBinding(qosConf, pod) && qos.IsPodNumaExclusive(qosConf, pod) && podNum > 1 {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+// checkNumaAllocatableSumAnomaly checks whether the node's sum of numa allocatable is not equal to the node allocatable
+func (ctrl *CNRMonitorController) checkNumaAllocatableSumAnomaly(cnr *v1alpha1.CustomNodeResource) bool {
+	// refer to https://github.com/kubewharf/katalyst-core/blob/main/pkg/agent/resourcemanager/fetcher/system/systemplugin.go#L126
+	// If the order of cpu and memory changes, the sequence number needs to be changed here as well
+	if len(cnr.Spec.NodeResourceProperties) < 3 {
+		klog.Errorf("[CNRIndicatorNumaAllocatableSumAnomaly] failed to get nodeResourceProperties")
+		return false
+	}
+	nodeCpuAllocatable, nodeMemAllocatable := int(cnr.Spec.NodeResourceProperties[1].PropertyQuantity.AsApproximateFloat64()), int(cnr.Spec.NodeResourceProperties[2].PropertyQuantity.AsApproximateFloat64())
```

Comment:
NodeCpuAllocatable and NodeMemAllocatable should be directly obtained from node status, not relying on NodeResourceProperties order. Future addition of more NodeResourceProperties may cause issues if not adhered to.